### PR TITLE
AutoComplete: Avoid Git for non tracked files

### PR DESCRIPTION
**Pull request by gerhardol**
## Proposed changes

Requirement for #9056 

AutoComplete in FormCommit tries to list all files that are in worktree/index and parse them, to find names of functions etc to use in the autocompletion.
(Extensions and regexp in AutoCompleteRegexes.txt)
First git-diff is tried (OK, but empty), but git-show will raise a git error status. Better to not call Git here than to handle the error in #9056 

### Before

image

### After

No Git calls

## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by mstv — APPROVED** [review]
Thank you!

File: GitCommands/Git/GitModule.cs [hunk 1]
```diff
@@ -2725,9 +2725,19 @@ public bool IsDirtyDir()
                 : null;
         }
 
+        /// <summary>
+        /// Get the file contents in HEAD for the file
+        /// </summary>
+        /// <param name="file">The Git status item</param>
+        /// <returns>async Task with contents</returns>
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2732]:
```suggestion
        /// <returns>An awaitable task with the requested file content.</returns>
```

File: GitCommands/Git/GitModule.cs [hunk 2]
```diff
@@ -2725,9 +2725,19 @@ public bool IsDirtyDir()
                 : null;
         }
 
+        /// <summary>
+        /// Get the file contents in HEAD for the file
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2729]: I'm struggling to parse the meaning
```suggestion
        /// Get the file content.
```

File: GitCommands/Git/GitModule.cs [hunk 3]
```diff
@@ -2725,9 +2725,19 @@ public bool IsDirtyDir()
...
+        /// </summary>
+        /// <param name="file">The Git status item</param>
+        /// <returns>async Task with contents</returns>
         public async Task<string?> GetFileContentsAsync(GitItemStatus file)
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2733]: This should be:
```suggestion
        public async Task<string?> GetFileContentAsync(GitItemStatus file)
```

File: GitCommands/Git/GitModule.cs [hunk 4]
```diff
@@ -2725,9 +2725,19 @@ public bool IsDirtyDir()
...
         {
             var contents = new StringBuilder();
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:2735]:
```suggestion
            var content = new StringBuilder();
```

**Review by RussKie — COMMENTED** [review]
👍

File: GitCommands/Git/GitModule.cs (hunk 2 above)
Comment (gerhardol) [reply GitCommands/Git/GitModule.cs:2729]: Get the file contents for the HEAD commit
```suggestion
        /// Get the file contents for the HEAD commit
```

File: GitCommands/Git/GitModule.cs (hunk 3 above)
Comment (gerhardol) [reply GitCommands/Git/GitModule.cs:2733]: I suggest to keep the s
https://english.stackexchange.com/questions/56831/file-content-vs-file-contents

File: GitCommands/Git/GitModule.cs (hunk 1 above)
Comment (gerhardol) [reply GitCommands/Git/GitModule.cs:2732]: As discussed below (currently) changed to
```suggestion
        /// <returns>An awaitable task with the requested file contents.</returns>
```

File: GitCommands/Git/GitModule.cs [hunk 5]
```diff
@@ -2725,10 +2725,20 @@ public bool IsDirtyDir()
...
         public async Task<string?> GetFileContentsAsync(GitItemStatus file)
         {
-            var contents = new StringBuilder();
+            if (!file.IsTracked || (file.Staged == StagedStatus.Index && file.IsNew))
+            {
+                // File is not in Git
+                return null;
+            }
 
+            var contents = new StringBuilder();
```
Comment (mstv) [inline GitCommands/Git/GitModule.cs:2741]:
```suggestion
            StringBuilder contents = new();
```

Comment (gerhardol) [reply GitCommands/Git/GitModule.cs]: We could change this globally too
(this was just a moved line, I do normally not updated unrelated changes to keep diffs reviewable.)
Updated, no other changes than this and the squash

**Comment by gerhardol** [comment]
@msftbot merge in 24 hours

File: GitCommands/Git/GitModule.cs [hunk 6]
```diff
@@ -2725,10 +2725,20 @@ public bool IsDirtyDir()
...
+        /// </summary>
+        /// <param name="file">The Git status item</param>
+        /// <returns>An awaitable task with the requested file contents.</returns>
         public async Task<string?> GetFileContentsAsync(GitItemStatus file)
         {
-            var contents = new StringBuilder();
+            if (!file.IsTracked || (file.Staged == StagedStatus.Index && file.IsNew))
+            {
+                // File is not in Git
+                return null;
```
Comment (ivangrek) [inline GitCommands/Git/GitModule.cs:2738]: Task is null there.

Comment (gerhardol) [reply GitCommands/Git/GitModule.cs:2738]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/async-return-types

The Task is async, not the same as not async.
The task is not null, it returns a null string.

**Pull request merged**